apps/src/studio/samBatLogic.js

```diff
@@ -0,0 +1,137 @@
+var codegen = require('../codegen');
+var api = require('./api');
+
+/**
+ * Interface for a set of custom game logic for playlab
+ * @param {Studio} studio Reference to global studio object
+ * @interface CustomGameLogic
+ */
+function CustomGameLogic(studio) {}
+
+/**
+ * Logic to be run once per playlab tick
+ *
+ * @function
+ * @name CustomGameLogic#onTick
+ */
+
+/**
+ * Custom logic for the Sam the Bat levels
+ * @constructor
+ * @implements CustomGameLogic
+ */
+var SamBatLogic = function (studio) {
+  this.studio_ = studio;
+  this.cached_ = {};
+  this.samIndex = 0;
+};
+
+SamBatLogic.prototype.onTick = function () {
+  this.sam = this.studio_.sprite[this.samIndex];
+  
+  // Move Sam with arrow keys
+  for (var key in this.studio_.keyState) {
+    if (this.studio_.keyState[key] === 'keydown') { 
+      switch (key) {
+        case '37':
+          this.updateSam_(Direction.WEST);
+          break;
+        case '38':
+          this.updateSam_(Direction.NORTH);
+          break;
+        case '39':
+          this.updateSam_(Direction.EAST);
+          break;
+        case '40':
+          this.updateSam_(Direction.SOUTH);
+          break;
+      }
+    }
+  }
+
+  // Move Sam with arrow buttons
+  for (var btn in this.studio_.btnState) {
+    if (this.studio_.btnState[btn]) {
+      switch (btn) {
+        case 'leftButton':
+          this.updateSam_(Direction.WEST);
+          break;
+        case 'upButton':
+          this.updateSam_(Direction.NORTH);
+          break;
+        case 'rightButton':
+          this.updateSam_(Direction.EAST);
+          break;
+        case 'downutton':
+          this.updateSam_(Direction.SOUTH);
+          break;
+      }
+    }
+  }
+  
+  // Display Sam's coordinates, with y inverted
+  var centerX = this.sam.x + this.sam.width / 2;
+  var centerY = this.studio_.MAZE_HEIGHT - (this.sam.y + this.sam.height / 2);
+  this.studio_.scoreText = '(' + centerX + ', ' + centerY + ')';
+  this.studio_.displayScore();
+};
+
+/**
+ * Before moving, check if Sam would still be onscreen?
+ * If move would take Sam offscreen, set dir to None
+ */
+SamBatLogic.prototype.updateSam_ = function (dir) {
+  var centerX = this.sam.x + this.sam.width / 2;
+  //invert Y
+  var centerY = this.studio_.MAZE_HEIGHT - (this.sam.y + this.sam.height / 2);
+  
+  switch (dir) {
+    case Direction.WEST:
+      if (!this.onscreen(centerX - this.sam.speed, centerY)) { dir = Direction.NONE; }
+      break;
+    case Direction.NORTH:
+      if (!this.onscreen(centerX, centerY + this.sam.speed)) { dir = Direction.NONE; }
+      break;
+    case Direction.EAST:
+      if (!this.onscreen(centerX + this.sam.speed, centerY)) { dir = Direction.NONE; }
+      break;
+    case Direction.SOUTH:
+      if (!this.onscreen(centerX, centerY - this.sam.speed)) { dir = Direction.NONE; }
+      break;
+  }
+  this.studio_.moveSingle({spriteIndex:this.samIndex, dir:dir});
+};
+
+SamBatLogic.prototype.cacheBlock = function (key, block) {
```

Comment:
This is fine for now, though we'd probably prefer to avoid duplicating this cache logic. FYI, I've opened a pivotal item assigned to me to figure out what can be extracted into a base class.